File: 1992-find-all-groups-of-farmland/1992-find-all-groups-of-farmland.py
class Solution:
    def findFarmland(self, land: List[List[int]]) -> List[List[int]]:
        def dfs(i, j, res):
            if (
                (j != n - 1 and land[i][j + 1] == 1) and
                (i != m - 1 and land[i + 1][j] == 1)
            ):
                dfs(i + 1, j + 1, res)
            elif (j != n - 1 and land[i][j + 1] == 1):
                dfs(i, j + 1, res)
            elif (i != m - 1 and land[i + 1][j] == 1):
                dfs(i + 1, j, res)
            else:
                res.extend([i, j])
                for i in range(res[0], res[-2] + 1):
                    for j in range(res[1], res[-1] + 1):
                        visited.add((i, j))

        ans = []

        m = len(land)
        n = len(land[0])
        
        visited = set()

        for i in range(m):
            for j in range(n):
                if land[i][j] and (i, j) not in visited:
                    res = [i, j]
                    visited.add((i, j))
                    dfs(i, j, res)
                    ans.append(res)

        return ans

File: 1992-find-all-groups-of-farmland/1992-find-all-groups-of-farmland.cpp
#include <vector>
#include <unordered_set>

class Solution {
public:
    std::vector<std::vector<int>> findFarmland(std::vector<std::vector<int>>& land) {
        std::vector<std::vector<int>> ans;
        int m = land.size();
        int n = land[0].size();
        std::unordered_set<std::string> visited;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (land[i][j] && visited.find(std::to_string(i) + "," + std::to_string(j)) == visited.end()) {
                    std::vector<int> res = {i, j};
                    visited.insert(std::to_string(i) + "," + std::to_string(j));
                    dfs(i, j, res, land, visited);
                    ans.push_back(res);
                }
            }
        }

        return ans;
    }

private:
    void dfs(int i, int j, std::vector<int>& res, std::vector<std::vector<int>>& land, std::unordered_set<std::string>& visited) {
        if ((j != land[0].size() - 1 && land[i][j + 1] == 1) && (i != land.size() - 1 && land[i + 1][j] == 1)) {
            dfs(i + 1, j + 1, res, land, visited);
        } else if (j != land[0].size() - 1 && land[i][j + 1] == 1) {
            dfs(i, j + 1, res, land, visited);
        } else if (i != land.size() - 1 && land[i + 1][j] == 1) {
            dfs(i + 1, j, res, land, visited);
        } else {
            res.push_back(i);
            res.push_back(j);
            for (int x = res[0]; x <= res[2]; x++) {
                for (int y = res[1]; y <= res[3]; y++) {
                    visited.insert(std::to_string(x) + "," + std::to_string(y));
                }
            }
        }
    }
};

